src/PinterestShippingZone.php

```diff
@@ -0,0 +1,261 @@
+<?php
+/**
+ * Represents a single Pinterest shipping zone
+ *
+ * @since   x.x.x
+ * @package Automattic\WooCommerce\Pinterest
+ */
+
+namespace Automattic\WooCommerce\Pinterest;
+
+defined( 'ABSPATH' ) || exit;
+
+use \WC_Shipping_Zone;
+
+/**
+ * PinterestShippingZone class.
+ *
+ * By extending WC_Shipping_Zone we are able to add functionality necessary for Pinterest shipping column generation.
+ * This allows us to operate on familiar interface which will be useful when we will continue to expand the functionality.
+ *
+ * @since x.x.x
+ */
+class PinterestShippingZone extends WC_Shipping_Zone {
+
+	/**
+	 * Caching for internal structure of locations.
+	 *
+	 * @var $zone_countries_with_states
+	 */
+	private $zone_countries_with_states = null;
+
+	/**
+	 * Caching for supported shipping methods.
+	 *
+	 * @var $zone_countries_with_states
+	 */
+	private $supported_shipping_methods = null;
+
+	/**
+	 * Types of allowed shipping methods.
+	 */
+	const ALLOWED_SHIPPING_METHODS = array( 'free_shipping', 'flat_rate' );
+
+	/**
+	 * Type of settings in the required field that we support.
+	 */
+	const ALLOWED_FREE_SHIPPING_REQUIRED_SETTINGS = array( '', 'min_amount' );
+
+	/**
+	 * From the list of countries filter out those which are not supported right now.
+	 *
+	 * @since   x.x.x
+	 *
+	 * @param  array $locations List of countries to filter.
+	 * @return array            List of filtered countries.
+	 */
+	private function filter_out_not_allowed_countries( $locations ) {
+		$allowed_countries = \Pinterest_For_Woocommerce_Ads_Supported_Countries::get_countries();
+		return array_filter(
+			$locations,
+			function ( $location ) use ( $allowed_countries ) {
+				return in_array( $location['country'], $allowed_countries, true );
+			}
+		);
+	}
+
+	/**
+	 * From the zone settings generate a list of countries with states.
+	 * States are optional and added only if specified.
+	 * Countries are filtered to only include supported countries.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return array Array of locations supported by this zone.
+	 */
+	public function get_countries_with_states() {
+		if ( null !== $this->zone_countries_with_states ) {
+			return $this->zone_countries_with_states;
+		}
+
+		$all_continents = WC()->countries->get_continents();
+		$zone_locations = $this->get_zone_locations();
+		$continents     = array_filter( $zone_locations, array( $this, 'location_is_continent' ) );
+		$countries      = array_filter( $zone_locations, array( $this, 'location_is_country' ) );
+		$states         = array_filter( $zone_locations, array( $this, 'location_is_state' ) );
+		$postcodes      = array_filter( $zone_locations, array( $this, 'location_is_postcode' ) );
```

Comment:
❓ 

Could we not group the zone locations like this: 

```
$type_areas = [];

foreach ( $zone_locations as $zone ) {
      $type_areas[ $zone->type ][] = $zone;
}
```

Instead of iterating the  `$zone_locations` for each type?

Reply:
We could. Both approaches would work.
I like the direct approach of having a separate variable for each type - this makes the later code easier to write.

What needs to happen in your example is that we need to prepopulate the `$type_areas` array with empty arrays for each type before we append any item. Or detect in the loop if the key exist and add it if not.

For me a flat structure is easier to grasp. Iterating over zone locations for each type is cheap so we don't have to worry about it.